File: android/app/src/main/java/com/albbiz/map/data/Story.kt
// Bismillah Hir Rahman Nir Raheem
package com.albbiz.map.data

data class Story(
    val id: String = "",
    val userId: String = "",
    val userName: String = "",
    val businessId: String? = null,
    val businessName: String? = null,
    val type: String = "user", // business, community, sponsored, user
    val category: String = "",
    val location: String = "",
    val photos: List<String> = emptyList(),
    val text: String = "",
    val createdAt: Long = System.currentTimeMillis(),
    val expiresAt: Long = System.currentTimeMillis() + 24 * 60 * 60 * 1000, // 24hrs
    val viewedBy: List<String> = emptyList(),
    val isSponsored: Boolean = false
) {
    fun toMap(): Map<String, Any?> {
        return mapOf(
            "id" to id,
            "userId" to userId,
            "userName" to userName,
            "businessId" to businessId,
            "businessName" to businessName,
            "type" to type,
            "category" to category,
            "location" to location,
            "photos" to photos,
            "text" to text,
            "createdAt" to createdAt,
            "expiresAt" to expiresAt,
            "viewedBy" to viewedBy,
            "isSponsored" to isSponsored
        )
    }

    companion object {
        fun fromMap(id: String, map: Map<String, Any?>): Story {
            return Story(
                id = id,
                userId = map["userId"] as? String ?: "",
                userName = map["userName"] as? String ?: "",
                businessId = map["businessId"] as? String,
                businessName = map["businessName"] as? String,
                type = map["type"] as? String ?: "user",
                category = map["category"] as? String ?: "",
                location = map["location"] as? String ?: "",
                photos = (map["photos"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                text = map["text"] as? String ?: "",
                createdAt = (map["createdAt"] as? Number)?.toLong() ?: System.currentTimeMillis(),
                expiresAt = (map["expiresAt"] as? Number)?.toLong() ?: System.currentTimeMillis() + 24 * 60 * 60 * 1000,
                viewedBy = (map["viewedBy"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
                isSponsored = map["isSponsored"] as? Boolean ?: false
            )
        }
    }
}

File: ios/AlbBizMap_IOS/AlbBizMap/Models/Story.swift
// Bismillah Hir Rahman Nir Raheem
import Foundation

public struct Story: Identifiable, Codable, Hashable {
    public var id: String
    public var userId: String
    public var userName: String
    public var businessId: String?
    public var businessName: String?
    public var type: String // business, community, sponsored, user
    public var category: String
    public var location: String
    public var photos: [String]
    public var text: String
    public var createdAt: Int64
    public var expiresAt: Int64
    public var viewedBy: [String]
    public var isSponsored: Bool
    
    public init(
        id: String = "",
        userId: String = "",
        userName: String = "",
        businessId: String? = nil,
        businessName: String? = nil,
        type: String = "user",
        category: String = "",
        location: String = "",
        photos: [String] = [],
        text: String = "",
        createdAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        expiresAt: Int64 = Int64(Date().timeIntervalSince1970 * 1000) + (24 * 60 * 60 * 1000),
        viewedBy: [String] = [],
        isSponsored: Bool = false
    ) {
        self.id = id
        self.userId = userId
        self.userName = userName
        self.businessId = businessId
        self.businessName = businessName
        self.type = type
        self.category = category
        self.location = location
        self.photos = photos
        self.text = text
        self.createdAt = createdAt
        self.expiresAt = expiresAt
        self.viewedBy = viewedBy
        self.isSponsored = isSponsored
    }
    
    public func toMap() -> [String: Any?] {
        return [
            "id": id,
            "userId": userId,
            "userName": userName,
            "businessId": businessId,
            "businessName": businessName,
            "type": type,
            "category": category,
            "location": location,
            "photos": photos,
            "text": text,
            "createdAt": createdAt,
            "expiresAt": expiresAt,
            "viewedBy": viewedBy,
            "isSponsored": isSponsored
        ]
    }
    
    public static func fromMap(id: String, map: [String: Any?]) -> Story {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        return Story(
            id: id,
            userId: map["userId"] as? String ?? "",
            userName: map["userName"] as? String ?? "",
            businessId: map["businessId"] as? String,
            businessName: map["businessName"] as? String,
            type: map["type"] as? String ?? "user",
            category: map["category"] as? String ?? "",
            location: map["location"] as? String ?? "",
            photos: (map["photos"] as? [String]) ?? [],
            text: map["text"] as? String ?? "",
            createdAt: (map["createdAt"] as? NSNumber)?.int64Value ?? now,
            expiresAt: (map["expiresAt"] as? NSNumber)?.int64Value ?? (now + 24 * 60 * 60 * 1000),
            viewedBy: (map["viewedBy"] as? [String]) ?? [],
            isSponsored: map["isSponsored"] as? Bool ?? false
        )
    }
}
